# Pandas

Import the libraries that are needed

In [1]:
import pandas as pd
import numpy as np

Pandas has two objects, namely series and data frames

# Object Series

Object Series has one data dimension, it does not have column names because it only has one column. And it has an index.

In [2]:
data = [0.25, 0.50, 0.75, 1]

In [3]:
print(data)

[0.25, 0.5, 0.75, 1]


convert data into series form 

In [4]:
data = pd.Series(data)

In [5]:
data

0    0.25
1    0.50
2    0.75
3    1.00
dtype: float64

Task 1 (create a list, convert it into series form)

In [6]:
import pandas as pd

data = ["senin", "selasa", "rabu", "kamis", "jumat"]

print(data)

['senin', 'selasa', 'rabu', 'kamis', 'jumat']


In [7]:
data = pd.Series(data)

data

0     senin
1    selasa
2      rabu
3     kamis
4     jumat
dtype: object

convert from series to array

In [8]:
data.values

array(['senin', 'selasa', 'rabu', 'kamis', 'jumat'], dtype=object)

displays index

The index is in the form of a range, where the start point is inclusive of the range and the stop point is exclusive of the range

In [9]:
data.index

RangeIndex(start=0, stop=5, step=1)

In [10]:
list(range(1,10))

[1, 2, 3, 4, 5, 6, 7, 8, 9]

how to call data

In [11]:
data[2]

'rabu'

Task 2 (convert series to array form, display index, slicing)

In [12]:
import pandas as pd

data = ["senin", "selasa", "rabu", "kamis", "jumat"]

print(data)

['senin', 'selasa', 'rabu', 'kamis', 'jumat']


In [13]:
data = pd.Series(data)

data

0     senin
1    selasa
2      rabu
3     kamis
4     jumat
dtype: object

In [14]:
data.values

array(['senin', 'selasa', 'rabu', 'kamis', 'jumat'], dtype=object)

In [15]:
data.index

RangeIndex(start=0, stop=5, step=1)

In [16]:
data[4]

'jumat'

The implicit index is the default index

we can define the index, this is called an explicit index, namely a defined index.

when defining indexes, the number of indexes must be equal to the number of data.

In [17]:
data = pd.Series([0.25, 0.50, 0.75, 1], index=['a', 'b', 'c', 'd'])

In [18]:
data

a    0.25
b    0.50
c    0.75
d    1.00
dtype: float64

In [19]:
data.values

array([0.25, 0.5 , 0.75, 1.  ])

In [20]:
data.index

Index(['a', 'b', 'c', 'd'], dtype='object')

calling the data

In [21]:
# indeks eksplisit

data['a']

0.25

This is data selection

Even though we have created an explicit index, we can still call the implicit index.

In [22]:
# indeks implisit

data[3]

1.0

Task 3 (create index explicit(defined index), display index, call index explicit, call index implisit)

In [23]:
data = pd.Series([0.25, 1, 1.75 , 2.25], index=['a', 'b', 'c', 'd'])
data

a    0.25
b    1.00
c    1.75
d    2.25
dtype: float64

In [24]:
data.values

array([0.25, 1.  , 1.75, 2.25])

In [25]:
data.index

Index(['a', 'b', 'c', 'd'], dtype='object')

In [26]:
# indeks eksplisit

data['b']

1.0

In [27]:
# indeks implisit

data[0]

0.25

when the implicit index and explicit index are the same, when we call the data, it will only duplicate the explicit index

In [28]:
data_2 = pd.Series([0.25, 0.50, 0.75, 1], index=[2,5,3,7])

In [29]:
data_2[2]

0.25

In [30]:
data_2[0]

KeyError: 0

Task 4 (create series dari list yang isinya 10 elemen, create index explicit (define index) yang ada kesamaan dengan index implicit, call index explicit and call index implicit)

In [31]:
data3 = pd.Series([3,4,5,6,7,8,9,10,11,12], index=[1,2,3,4,5,6,7,8,9,10])
data3

1      3
2      4
3      5
4      6
5      7
6      8
7      9
8     10
9     11
10    12
dtype: int64

In [32]:
data.values

array([0.25, 1.  , 1.75, 2.25])

In [33]:
data.index

Index(['a', 'b', 'c', 'd'], dtype='object')

In [34]:
data3[3]

5

In [35]:
data3[4]

6

we will try to do data slicing

In [36]:
data = pd.Series([0.25, 1, 1.75 , 2.25], index=['a', 'b', 'c', 'd'])

In [37]:
data

a    0.25
b    1.00
c    1.75
d    2.25
dtype: float64

In [38]:
data['b':'c'] # indeks eksplisit

b    1.00
c    1.75
dtype: float64

but if we slice the implicit index, only the starting point will appear, because the implicit index is a range

In [39]:
data[1:2] # indeks implisit

b    1.0
dtype: float64

In [40]:
data_2[2:3] #indeks implisit : slicing

3    0.75
dtype: float64

Task 5 (create index explicit, slicing explicit, slicing implicit with two params and 3 params)

In [41]:
data = pd.Series([1,2,3,4,5,6,7,8,9,10], index=['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j'])

In [42]:
data

a     1
b     2
c     3
d     4
e     5
f     6
g     7
h     8
i     9
j    10
dtype: int64

In [43]:
data['a':'j'] # indeks eksplisit

a     1
b     2
c     3
d     4
e     5
f     6
g     7
h     8
i     9
j    10
dtype: int64

In [44]:
data[1:7] # indeks implisit 2 parameter

b    2
c    3
d    4
e    5
f    6
g    7
dtype: int64

In [45]:
data[1:7:2] # indeks implisit 3 parameter

b    2
d    4
f    6
dtype: int64

# LOC AND ILOC

Loc is for calling its explicit index.

Iloc is for calling its implicit index.

In [46]:
#loc

data_2.loc[3] #selecting indeks eksplisit

0.75

In [47]:
data_2.loc[2:3] #slicing indeks eksplisit 

2    0.25
5    0.50
3    0.75
dtype: float64

In [48]:
#iloc


data_2.iloc[3] #selecting indeks implisit

1.0

In [49]:
data_2.iloc[2:3] #slicing indeks implisit

3    0.75
dtype: float64

Task 1 (create series, create indeks eksplisit with same index in python, call loc and iloc)

In [50]:
data = pd.Series([1,2,3,4,5,6,7,8,9,10], index=['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j'])
data

a     1
b     2
c     3
d     4
e     5
f     6
g     7
h     8
i     9
j    10
dtype: int64

In [51]:
data['a':'j'] # indeks eksplisit

a     1
b     2
c     3
d     4
e     5
f     6
g     7
h     8
i     9
j    10
dtype: int64

In [52]:
#loc

data.loc['c']

3

In [53]:
#iloc

data.iloc[2]

3

In [54]:
dict_populasi = {'Jakarta':750,
                 'Bogor':400,
                 'Depok':350,
                 'Tanggerang':270,
                 'Bekasi':670}

In [55]:
dict_populasi

{'Jakarta': 750, 'Bogor': 400, 'Depok': 350, 'Tanggerang': 270, 'Bekasi': 670}

In [56]:
#transformasi dictionary ke series

populasi = pd.Series(dict_populasi)

In [57]:
populasi

Jakarta       750
Bogor         400
Depok         350
Tanggerang    270
Bekasi        670
dtype: int64

In [58]:
populasi.loc['Depok']

350

In [59]:
populasi.iloc[2]

350

Task 2 (create series from dictionary, call loc, call iloc)

In [60]:
kelompok_5 = {'Anggota1':'Renata',
              'Anggota2':'Mala',
             'Anggota3':'Ardi',
             'Anggota4':'Novan'}

In [61]:
kelompok_5

{'Anggota1': 'Renata',
 'Anggota2': 'Mala',
 'Anggota3': 'Ardi',
 'Anggota4': 'Novan'}

In [62]:
kelompok = pd.Series(kelompok_5)

In [63]:
kelompok

Anggota1    Renata
Anggota2      Mala
Anggota3      Ardi
Anggota4     Novan
dtype: object

In [64]:
kelompok.loc['Anggota1']

'Renata'

In [65]:
kelompok.iloc[0]

'Renata'

In [66]:
dict_luas = {'Jakarta':737,
                 'Bogor':325,
                 'Depok':247,
                 'Tanggerang':302,
                 'Bekasi':355}

In [67]:
luas = pd.Series(dict_luas)

In [68]:
luas

Jakarta       737
Bogor         325
Depok         247
Tanggerang    302
Bekasi        355
dtype: int64

# Data Frame

Data Frame is a collection of series, with at least one series

In [69]:
daerah = pd.DataFrame({'pop':populasi, 'luas daerah':luas})

In [70]:
daerah

pop  luas daerah
Jakarta     750          737
Bogor       400          325
Depok       350          247
Tanggerang  270          302
Bekasi      670          355

In [71]:
daerah['luas daerah']

Jakarta       737
Bogor         325
Depok         247
Tanggerang    302
Bekasi        355
Name: luas daerah, dtype: int64

In [72]:
daerah['luas daerah']['Jakarta']

737

In [73]:
daerah = pd.Series({'populasi':populasi, 'luas':luas})

In [74]:
daerah

populasi    Jakarta       750
Bogor         400
Depok     ...
luas        Jakarta       737
Bogor         325
Depok     ...
dtype: object

In [75]:
daerah['populasi']

Jakarta       750
Bogor         400
Depok         350
Tanggerang    270
Bekasi        670
dtype: int64

In [76]:
daerah['luas daerah']['Jakarta':'Depok'] #index eksplisit

KeyError: 'luas daerah'

In [77]:
daerah['populasi'].iloc[0:9] #index implisit

Jakarta       750
Bogor         400
Depok         350
Tanggerang    270
Bekasi        670
dtype: int64

Task 3 (create 3 series, replace into a data frame, call index)

In [78]:
kelompok_5 = {'Anggota1':'Renata',
              'Anggota2':'Mala',
             'Anggota3':'Ardi',
             'Anggota4':'Novan'}

In [79]:
kul_semester = {'Anggota1':'5',
              'Anggota2':'5',
             'Anggota3':'5',
             'Anggota4':'5'}

In [80]:
asal_daerah = {'Anggota1':'Malang',
              'Anggota2':'Jakarta',
             'Anggota3':'Lamongan',
             'Anggota4':'Jepara'}

In [81]:
data_kel5 = pd.DataFrame({'Nama':kelompok_5, 'Semester':kul_semester, 'Domisili':asal_daerah})

In [82]:
data_kel5['Nama']

Anggota1    Renata
Anggota2      Mala
Anggota3      Ardi
Anggota4     Novan
Name: Nama, dtype: object

In [83]:
data_kel5['Semester']['Anggota1']

'5'

In [84]:
data_kel5

Nama Semester  Domisili
Anggota1  Renata        5    Malang
Anggota2    Mala        5   Jakarta
Anggota3    Ardi        5  Lamongan
Anggota4   Novan        5    Jepara

Task 4 (replace column name, call loc and iloc)

In [85]:
data_kel5 = pd.DataFrame({'Nama Anggota':kelompok_5, 'Kuliah SMT':kul_semester, 'Asal Daerah':asal_daerah})

In [86]:
data_kel5

Nama Anggota Kuliah SMT Asal Daerah
Anggota1       Renata          5      Malang
Anggota2         Mala          5     Jakarta
Anggota3         Ardi          5    Lamongan
Anggota4        Novan          5      Jepara

In [87]:
data_kel5['Nama Anggota']['Anggota1':'Anggota3']

Anggota1    Renata
Anggota2      Mala
Anggota3      Ardi
Name: Nama Anggota, dtype: object

In [88]:
data_kel5['Nama Anggota'].iloc[0:3]

Anggota1    Renata
Anggota2      Mala
Anggota3      Ardi
Name: Nama Anggota, dtype: object

# Titanic csv

In [89]:
#kita load data titanic

df = pd.read_csv('Titanic.csv')

In [90]:
#Displays Data From Head

df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [91]:
#Displays Info Data

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Task 1 (Displays the top 10 data and the bottom 10 data, show which columns contain missing values, show columns whose data types are incomplete)

In [92]:
import pandas as pd

In [93]:
df = pd.read_csv('Titanic.csv')

In [94]:
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [95]:
df.tail(10)

PassengerId  Survived  Pclass                                      Name  \
881          882         0       3                        Markun, Mr. Johann   
882          883         0       3              Dahlberg, Miss. Gerda Ulrika   
883          884         0       2             Banfield, Mr. Frederick James   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
881    male  33.0      0      0            349257   7.8958   NaN        S  
882  female  22.0      0      0              7552  10.5167   NaN        S  
883    male  28.0      0      0  C.A./SOTON 34068  10.5000   NaN        S  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
886    male  27.0      0      0            211536  13.0000   NaN        S  
887  female  19.0      0      0            112053  30.0000   B42        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
889    male  26.0      0      0            111369  30.0000  C148        C  
890    male  32.0      0      0            370376   7.7500   NaN        Q

answer = Columns with missing values: Embarked, Cabin, and Age. This is because there are values ​​containing NaN in that column.

In [96]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


answer = The column that is not appropriate is the Age column. Because in general age calculations only use the year of birth.

In [97]:
#Displays the number of rows and the number of columns

df.shape

(891, 12)

In [98]:
#Displays Data From columns

df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [99]:
#Displays Data Fromindex

df.index

RangeIndex(start=0, stop=891, step=1)

In [100]:
#Displays Information From Columns in The From of Number

df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [101]:
#Displays The Sum Of Not Null From Head

df.notnull().sum()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

In [102]:
#Displays The Sum Is Null From Head

df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [103]:
#Displays The Amount From Head

df.sum

<bound method NDFrame._add_numeric_operations.<locals>.sum of      PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..     

Task 2 (dig for insights other than column age)

In [104]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

##### Insight Kolom Fare
* Harga tiket paling mahal adalah 512.3292
* Harga tiket paling murah adalah 0 (asumsi mata uangnya adalah dollar)
* Rata-rata harga tiketnya adalah 32.204208
* Semua penumpang kapal membeli tiket karena jumlah count baris sama dengan range index

##### Insight Kolom Survived
* Rata-rata kemungkinan hidup penumpang titanic adalah 0,38 atau 38%
* Semua orang dapat diidentifikasi status keselamatannya (masih hidup atau sudah meninggal)

In [105]:
import pandas as pd

In [106]:
df = pd.read_csv('Titanic.csv')

In [107]:
#Displays The Mean of Columns

df['Age'].mean()

29.69911764705882

In [108]:
#Displays The Median of Columns

df['Age'].median()

28.0

In [109]:
#Displays The Modus of Columns


df['Age'].mode()

0    24.0
Name: Age, dtype: float64

In [110]:
#Displays The Minimal of Columns

df['Age'].min()

0.42

In [111]:
#Displays The Maximal of Columns

df['Age'].max()

80.0